search-books/src/main/java/nix/education/service/query/solr/SolrQueryBuilderServiceImpl.java

```diff
@@ -0,0 +1,71 @@
+package nix.education.service.query.solr;
+
+import java.util.Date;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.StringJoiner;
+import nix.education.util.DateFormat;
+import nix.education.util.SearchConstants;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+@Service
+public class SolrQueryBuilderServiceImpl implements SolrQueryBuilderService {
+
+    @Value("${search.term-facets}")
+    private Set<String> termFacets;
+
+    @Value("${search.range-facets}")
+    private Set<String> rangeFacets;
+
+    @Value("#{${search.search-fields}}")
+    private Map<String, Float> fields;
+
+    @Override
+    public SolrQuery getQuery(String q, Optional<Set<String>> fq) {
+        SolrQuery query = new SolrQuery();
+        query.add(SearchConstants.QUERY, buildSearchQuery(q));
+        termFacets.forEach(query::addFacetField);
+        query.addDateRangeFacet("publication_date", DateFormat.getStartSearchBookDate(),
+            new Date(System.currentTimeMillis()),
+            String.format("+%dYEAR", SearchConstants.DATE_RANGE_FACET_GAP));
+        addFiltersToQuery(query, fq);
+
+        return query;
+    }
+
+    private String buildSearchQuery(String q) {
+        StringJoiner stringJoiner = new StringJoiner(SearchConstants.OR);
+        for (Map.Entry<String, Float> field : fields.entrySet()) {
+            stringJoiner.add(field.getKey() +
+                SearchConstants.COLON + q + SearchConstants.CIRCUMFLEX + field.getValue());
```

Comment:
Multiple string concatenation doesn't look well. May be try this way:
`StringUtils.join(field.getKey(), SearchConstants.COLON, q, SearchConstants.CIRCUMFLEX, field.getValue())`